File: packages/python/taiwan_payroll/engine/prorated.py
from __future__ import annotations

import re

from .._lookup import find_bracket, brackets_for
from .._rounding import apply_rate_prorated, number_to_rate_string
from .._identity import get_identity_rules
from .._types import ProratedInput, ProratedResult, Brackets, Employee, Employer, Government, Meta, Days
from .health import calc_health_insurance

_ISO = re.compile(r"^\d{4}-\d{2}-\d{2}$")


def _clamp(n: int, lo: int, hi: int) -> int:
    return max(lo, min(hi, n))


def _day(d: str) -> int:
    return int(d[8:10])


def _ym(d: str) -> str:
    return d[0:7]


def _assert_iso(d: str) -> None:
    if not _ISO.match(d):
        raise ValueError(f'Date must be in YYYY-MM-DD format, got "{d}"')


def compute_insured_days(start_date: str | None, end_date: str | None) -> int:
    if not start_date and not end_date:
        raise ValueError("calculateProrated requires startDate or endDate")
    if start_date:
        _assert_iso(start_date)
    if end_date:
        _assert_iso(end_date)
    if start_date and end_date and _ym(start_date) != _ym(end_date):
        raise ValueError("startDate and endDate must be in the same month")
    if start_date and end_date:
        s = min(_day(start_date), 30)
        e = min(_day(end_date), 30)
        return _clamp(e - s + 1, 0, 30)
    if start_date:
        return _clamp(30 - min(_day(start_date), 30) + 1, 0, 30)
    return _clamp(min(_day(end_date), 30), 0, 30)


def health_charged_this_month(start_date: str | None, end_date: str | None) -> bool:
    return not end_date


def calc_prorated(data: dict, inp: ProratedInput, occupational_rate: str) -> ProratedResult:
    days = compute_insured_days(inp.start_date, inp.end_date)
    health_charged = health_charged_this_month(inp.start_date, inp.end_date)
    rounding = inp.rounding
    part_time = inp.part_time
    salary = inp.monthly_salary
    rules = get_identity_rules(inp.identity)
    employment_insurance = inp.employment_insurance if rules.employment_insurance_allowed else False

    li = data["laborInsurance"]
    labor_insured = labor_employee = labor_employer = labor_gov = 0
    if rules.labor_applies:
        labor_insured = find_bracket(brackets_for(li["brackets"], li.get("partTimeBrackets"), part_time), salary)["insuredSalary"]
        labor_rate = li["rate"] if employment_insurance else li["rateWithoutEmploymentInsurance"]
        b = li["burden"]
        if rounding == "aggregate-then-round":
            labor_total = apply_rate_prorated(labor_insured, [labor_rate], days, 30, "round")
            labor_employee = apply_rate_prorated(labor_insured, [labor_rate, b["employee"]], days, 30, "round")
            labor_employer = apply_rate_prorated(labor_insured, [labor_rate, b["employer"]], days, 30, "round")
            labor_gov = labor_total - labor_employee - labor_employer
        else:
            labor_employee = apply_rate_prorated(labor_insured, [labor_rate, b["employee"]], days, 30, rounding)
            labor_employer = apply_rate_prorated(labor_insured, [labor_rate, b["employer"]], days, 30, rounding)
            labor_gov = apply_rate_prorated(labor_insured, [labor_rate, b["government"]], days, 30, rounding)

    oi = data["occupationalInsurance"]
    occ_insured = find_bracket(oi["brackets"], salary)["insuredSalary"]
    occ_employer = apply_rate_prorated(occ_insured, [occupational_rate], days, 30, rounding)

    p = data["pension"]
    pen_insured = pen_employer = pen_self = 0
    if rules.pension_applies:
        pen_insured = find_bracket(p["brackets"], salary)["insuredSalary"]
        pen_employer = apply_rate_prorated(pen_insured, [p["employerRate"]], days, 30, rounding)
        clamped = min(max(inp.pension_self_contribution, 0.0), 0.06)
        pen_self = (
            apply_rate_prorated(pen_insured, [number_to_rate_string(clamped)], days, 30, rounding) if clamped > 0 else 0
        )

    hi = data["healthInsurance"]
    health_insured = find_bracket(brackets_for(hi["brackets"], hi.get("partTimeBrackets"), part_time), salary)["insuredSalary"]
    if health_charged:
        h = calc_health_insurance(data, salary, inp.dependents, rounding, part_time)
    else:
        h = {"insured": health_insured, "employee": 0, "employer": 0, "government": 0}

    return ProratedResult(
        brackets=Brackets(labor_insured, health_insured, pen_insured, occ_insured),
        employee=Employee(labor_employee, h["employee"], pen_self, labor_employee + h["employee"] + pen_self),
        employer=Employer(
            labor_employer,
            h["employer"],
            pen_employer,
            occ_employer,
            labor_employer + h["employer"] + pen_employer + occ_employer,
        ),
        government=Government(labor_gov, h["government"]),
        meta=Meta(data["year"], data["dataVersion"]),
        days=Days(days),
        health_charged=health_charged,
    )

File: packages/core/src/engine/prorated.ts
import type { YearData, ProratedInput, ProratedResult } from '../types';
import { findBracket, bracketsFor } from '../lookup/bracket';
import { applyRateProrated, numberToRateString } from '../rounding/strategies';
import { getIdentityRules } from '../identity';
import { calcHealthInsurance } from './healthInsurance';

const clamp = (n: number, lo: number, hi: number): number => Math.max(lo, Math.min(hi, n));
const ISO_DATE = /^\d{4}-\d{2}-\d{2}$/;
const dayOfMonth = (d: string): number => parseInt(d.slice(8, 10), 10);
const yearMonth = (d: string): string => d.slice(0, 7);

function assertIsoDate(d: string): void {
  if (!ISO_DATE.test(d)) throw new Error(`Date must be in YYYY-MM-DD format, got "${d}"`);
}

/** Insured days (30-day basis) for 勞保/職保/勞退. Throws on no date / cross-month range / malformed date. */
export function computeInsuredDays(startDate?: string, endDate?: string): number {
  if (!startDate && !endDate) throw new Error('calculateProrated requires startDate or endDate');
  if (startDate) assertIsoDate(startDate);
  if (endDate) assertIsoDate(endDate);
  if (startDate && endDate && yearMonth(startDate) !== yearMonth(endDate)) {
    throw new Error('startDate and endDate must be in the same month');
  }
  if (startDate && endDate) {
    const s = Math.min(dayOfMonth(startDate), 30);
    const e = Math.min(dayOfMonth(endDate), 30);
    return clamp(e - s + 1, 0, 30);
  }
  if (startDate) return clamp(30 - Math.min(dayOfMonth(startDate), 30) + 1, 0, 30);
  return clamp(Math.min(dayOfMonth(endDate as string), 30), 0, 30);
}

/** 健保 月底歸屬: charged full month unless the employee leaves this month. */
export function healthChargedThisMonth(_startDate?: string, endDate?: string): boolean {
  return !endDate;
}

export function calcProrated(data: YearData, input: ProratedInput, occupationalRate: string): ProratedResult {
  const days = computeInsuredDays(input.startDate, input.endDate);
  const healthCharged = healthChargedThisMonth(input.startDate, input.endDate);
  const rounding = input.rounding ?? 'round';
  const rules = getIdentityRules(input.identity ?? 'category1');
  const employmentInsurance = rules.employmentInsuranceAllowed ? (input.employmentInsurance ?? true) : false;
  const dependents = input.dependents ?? 0;
  const selfRate = input.pensionSelfContribution ?? 0;
  const partTime = input.partTime ?? false;
  const salary = input.monthlySalary;

  const li = data.laborInsurance;
  let laborInsured = 0;
  let laborEmployee = 0;
  let laborEmployer = 0;
  let laborGov = 0;
  if (rules.laborApplies) {
    laborInsured = findBracket(bracketsFor(li.brackets, li.partTimeBrackets, partTime), salary).insuredSalary;
    const laborRate = employmentInsurance ? li.rate : li.rateWithoutEmploymentInsurance;
    if (rounding === 'aggregate-then-round') {
      // Mirror calcLaborInsurance: round the prorated total once, government absorbs the split remainder.
      const laborTotal = applyRateProrated(laborInsured, [laborRate], days, 30, 'round');
      laborEmployee = applyRateProrated(laborInsured, [laborRate, li.burden.employee], days, 30, 'round');
      laborEmployer = applyRateProrated(laborInsured, [laborRate, li.burden.employer], days, 30, 'round');
      laborGov = laborTotal - laborEmployee - laborEmployer;
    } else {
      laborEmployee = applyRateProrated(laborInsured, [laborRate, li.burden.employee], days, 30, rounding);
      laborEmployer = applyRateProrated(laborInsured, [laborRate, li.burden.employer], days, 30, rounding);
      laborGov = applyRateProrated(laborInsured, [laborRate, li.burden.government], days, 30, rounding);
    }
  }

  const occInsured = findBracket(data.occupationalInsurance.brackets, salary).insuredSalary;
  const occEmployer = applyRateProrated(occInsured, [occupationalRate], days, 30, rounding);

  let penInsured = 0;
  let penEmployer = 0;
  let penSelf = 0;
  if (rules.pensionApplies) {
    penInsured = findBracket(data.pension.brackets, salary).insuredSalary;
    penEmployer = applyRateProrated(penInsured, [data.pension.employerRate], days, 30, rounding);
    const clampedSelf = Math.min(Math.max(selfRate, 0), 0.06);
    penSelf = clampedSelf > 0 ? applyRateProrated(penInsured, [numberToRateString(clampedSelf)], days, 30, rounding) : 0;
  }

  const healthInsured = findBracket(
    bracketsFor(data.healthInsurance.brackets, data.healthInsurance.partTimeBrackets, partTime),
    salary,
  ).insuredSalary;
  const health = healthCharged
    ? calcHealthInsurance(data, salary, dependents, rounding, partTime)
    : { insured: healthInsured, employee: 0, employer: 0, government: 0 };

  return {
    brackets: { labor: laborInsured, health: healthInsured, pension: penInsured, occupational: occInsured },
    employee: {
      labor: laborEmployee,
      health: health.employee,
      pensionSelf: penSelf,
      total: laborEmployee + health.employee + penSelf,
    },
    employer: {
      labor: laborEmployer,
      health: health.employer,
      pension: penEmployer,
      occupational: occEmployer,
      total: laborEmployer + health.employer + penEmployer + occEmployer,
    },
    government: { labor: laborGov, health: health.government },
    meta: { year: data.year, dataVersion: data.dataVersion },
    days: { insured: days },
    healthCharged,
  };
}
